api/clients/eigenda_client.go

```diff
@@ -0,0 +1,187 @@
+	case result := <-resultChan:
+		return result, nil
+	case err := <-errorChan:
+		return nil, err
+	}
+}
+
+func (m EigenDAClient) PutBlobAsync(ctx context.Context, data []byte) (resultChan chan *Cert, errChan chan error) {
+	resultChan = make(chan *Cert, 1)
+	errChan = make(chan error, 1)
+	go m.putBlob(ctx, data, resultChan, errChan)
+	return
+}
+
+func (m EigenDAClient) putBlob(ctx context.Context, rawData []byte, resultChan chan *Cert, errChan chan error) {
+	m.Log.Info("Attempting to disperse blob to EigenDA")
+
+	// encode blob
+	if m.PutCodec == nil {
+		errChan <- fmt.Errorf("PutCodec cannot be nil")
+		return
+	}
+	data := m.PutCodec.EncodeBlob(rawData)
+
+	customQuorumNumbers := make([]uint8, len(m.Config.CustomQuorumIDs))
+	for i, e := range m.Config.CustomQuorumIDs {
+		customQuorumNumbers[i] = uint8(e)
+	}
+
+	// disperse blob
+	blobStatus, requestID, err := m.Client.DisperseBlobAuthenticated(ctx, data, customQuorumNumbers)
+	if err != nil {
+		errChan <- fmt.Errorf("error initializing DisperseBlobAuthenticated() client: %w", err)
+		return
+	}
+
+	// process response
+	if *blobStatus == disperser.Failed {
+		m.Log.Error("Unable to disperse blob to EigenDA, aborting", "err", err)
+		errChan <- fmt.Errorf("reply status is %d", blobStatus)
+		return
+	}
+
+	base64RequestID := base64.StdEncoding.EncodeToString(requestID)
+	m.Log.Info("Blob dispersed to EigenDA, now waiting for confirmation", "requestID", base64RequestID)
+
+	ticker := time.NewTicker(m.Config.StatusQueryRetryInterval)
+	defer ticker.Stop()
+
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithTimeout(ctx, m.Config.StatusQueryTimeout)
+	defer cancel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			errChan <- fmt.Errorf("timed out waiting for EigenDA blob to confirm blob with request id=%s: %w", base64RequestID, ctx.Err())
+			return
+		case <-ticker.C:
+			statusRes, err := m.Client.GetBlobStatus(ctx, requestID)
+			if err != nil {
+				m.Log.Error("Unable to retrieve blob dispersal status, will retry", "requestID", base64RequestID, "err", err)
+				continue
+			}
+
+			switch statusRes.Status {
+			case grpcdisperser.BlobStatus_PROCESSING, grpcdisperser.BlobStatus_DISPERSING:
+				m.Log.Info("Blob submitted, waiting for dispersal from EigenDA", "requestID", base64RequestID)
+			case grpcdisperser.BlobStatus_FAILED:
+				m.Log.Error("EigenDA blob dispersal failed in processing", "requestID", base64RequestID, "err", err)
+				errChan <- fmt.Errorf("EigenDA blob dispersal failed in processing, requestID=%s: %w", base64RequestID, err)
+				return
+			case grpcdisperser.BlobStatus_INSUFFICIENT_SIGNATURES:
+				m.Log.Error("EigenDA blob dispersal failed in processing with insufficient signatures", "requestID", base64RequestID, "err", err)
+				errChan <- fmt.Errorf("EigenDA blob dispersal failed in processing with insufficient signatures, requestID=%s: %w", base64RequestID, err)
+				return
+			case grpcdisperser.BlobStatus_CONFIRMED:
```

Comment:
Couldn't there be a situation where a user wants to make their own determination of finalization status and to access the confirmation data in the meantime? Would they use the low-level client for this?

Reply:
Yes, the low level client is a way to do this for clients who don't want to wait for finalization, and PutAsync() exists on the high-level client if the client wants to avoid a long block, but still wants to wait for finalization.

In general though until we have batch confirmation retries we want clients to wait until batches are finalized, otherwise reorgs can break rollups. The easiest way to do this is to this is to encode the opinion in the client.

Reply:
> Couldn't there be a situation where a user wants to make their own determination of finalization status and to access the confirmation data in the meantime?

The only situation where the user would want to wait for confirmation but not finalization is if they're using some separate system for waiting for DA cert finalization, is there a specific use case you're thinking of?